### PR TITLE
Global context autocompletion in the Ivy Language Service

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

**Comment by alxhub** [comment]
Reviewers: only the last commit is relevant, the others are part of #39594 and will go away after rebase.

File: packages/language-service/ivy/completions.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import {NgCompiler} from '@angular/compiler-cli/src/ngtsc/core';
+import {CompletionKind, ReferenceSymbol, VariableSymbol} from '@angular/compiler-cli/src/ngtsc/typecheck/api';
+import * as ts from 'typescript';
+
+import {DisplayInfoKind, getDisplayInfo, unsafeCastDisplayInfoKindToScriptElementKind} from './display_parts';
+
+type PropertyExpressionCompletionBuilder = CompletionBuilder<PropertyRead|MethodCall>;
+
+/**
+ * Closure around the context required to perform autocompletion operations regarding a given node
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:19]: I think this would be better phrased as "Class providing a closure..." or "Class for building autocompletion... ", as the class itself is not strictly a closure.

File: packages/language-service/ivy/completions.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+      private readonly tsLS: ts.LanguageService, private readonly compiler: NgCompiler,
+      private readonly component: ts.ClassDeclaration, private readonly node: N,
+      private readonly context: TmplAstTemplate|null) {}
+
+  /**
+   * Analogue for `ts.LanguageService.getCompletionsAtPosition`.
+   */
+  getCompletionsAtPosition(options: ts.GetCompletionsAtPositionOptions|
+                           undefined): ts.WithMetadata<ts.CompletionInfo>|undefined {
+    console.error('getCompletionsAtPosition()');
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:41]:
```suggestion
```

File: packages/language-service/ivy/completions.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    }
+  }
+
+  /**
+   * Analogue for `ts.LanguageService.getCompletionEntryDetails`.
+   */
+  getCompletionEntryDetails(
+      entryName: string, formatOptions: ts.FormatCodeOptions|ts.FormatCodeSettings|undefined,
+      preferences: ts.UserPreferences|undefined): ts.CompletionEntryDetails|undefined {
+    console.error('getCompletionEntryDetails()', entryName);
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:55]:
```suggestion
```

File: packages/language-service/ivy/completions.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+  /**
+   * Analogue for `ts.LanguageService.getCompletionsAtPosition`.
+   */
+  getCompletionsAtPosition(options: ts.GetCompletionsAtPositionOptions|
+                           undefined): ts.WithMetadata<ts.CompletionInfo>|undefined {
+    console.error('getCompletionsAtPosition()');
+    if (this.isPropertyExpressionCompletion()) {
+      return this.getPropertyExpressionCompletion(options);
+    } else {
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:44]: nit, no action required: branch is unnecessary

File: packages/language-service/ivy/completions.ts [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+      return undefined;
+    }
+  }
+
+  /**
+   * Analogue for `ts.LanguageService.getCompletionEntrySymbol`.
+   */
+  getCompletionEntrySymbol(name: string): ts.Symbol|undefined {
+    console.error('getCompletionEntrySymbol()', name);
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:67]:
```suggestion
```

File: packages/language-service/ivy/completions.ts [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+  /**
+   * Analogue for `ts.LanguageService.getCompletionsAtPosition`.
+   */
+  getCompletionsAtPosition(options: ts.GetCompletionsAtPositionOptions|
+                           undefined): ts.WithMetadata<ts.CompletionInfo>|undefined {
+    console.error('getCompletionsAtPosition()');
+    if (this.isPropertyExpressionCompletion()) {
+      return this.getPropertyExpressionCompletion(options);
+    } else {
+      return undefined;
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:45]: Should there be a TODO/note here for other node kinds?

File: packages/language-service/ivy/completions.ts [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+  private isPropertyExpressionCompletion(this: CompletionBuilder<TmplAstNode|AST>):
+      this is PropertyExpressionCompletionBuilder {
+    return this.node instanceof PropertyRead || this.node instanceof MethodCall;
+  }
+
+  /**
+   * Get completions for property expressions.
+   */
+  private getPropertyExpressionCompletion(
+      this: PropertyExpressionCompletionBuilder,
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:91]: Rather than narrowing the type of `this`, could we just narrow the type of the node to retrieve completions for and pass it down as a parameter? To me, that would help avoid an extra context switch when reading and would flow nicer because we're getting the completions for a specific node rather than the class instance, but I can understand the motivation of the current pattern as well.

File: packages/language-service/ivy/completions.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+      }
+    }
+
+    for (const [name, entity] of templateContext) {
+      entries.push({
+        name,
+        sortText: name,
+        replacementSpan,
+        kindModifiers: ts.ScriptElementKindModifier.none,
+        kind: unsafeCastDisplayInfoKindToScriptElementKind(
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:176]: nit: add a comment about why this is cast is sound here

File: packages/language-service/ivy/completions.ts [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+            entity.kind === CompletionKind.Reference ? DisplayInfoKind.REFERENCE :
+                                                       DisplayInfoKind.VARIABLE),
+      });
+    }
+
+    return {
+      entries,
+      isGlobalCompletion: false,
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:184]: nit: add a comment why `isGlobalCompletion` is false, the function name `getGlobalPropertyExpressionCompletion` may suggest otherwise

File: packages/language-service/ivy/completions.ts [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+    if (templateContext.has(entryName)) {
+      const entry = templateContext.get(entryName)!;
+      const symbol =
+          templateTypeChecker.getSymbolOfNode(entry.node, this.component) as ReferenceSymbol |
+          VariableSymbol;
+      const {kind, displayParts, documentation} =
+          getDisplayInfo(this.tsLS, this.typeChecker, symbol);
+      return {
+        kind: kind as string as ts.ScriptElementKind,
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:213]: Should unsafeCastDisplayInfoKindToScriptElementKind be used here?

File: packages/language-service/ivy/completions.ts [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+    if (completions === null) {
+      return undefined;
+    }
+    const {componentContext, templateContext} = completions;
+
+    if (templateContext.has(entryName)) {
+      const entry = templateContext.get(entryName)!;
+      const symbol =
+          templateTypeChecker.getSymbolOfNode(entry.node, this.component) as ReferenceSymbol |
+          VariableSymbol;
```
Comment (ayazhafiz) [inline packages/language-service/ivy/completions.ts:209]: Could we pull out `ReferenceSymbol | VariableSymbol` as a type like `ValidTemplateSymbol` (I'm sure there is a much better name than that :) ); the motivation is to say that we are doing this because we know it can only be a ReferenceSymbol or VariableSymbol.

File: packages/language-service/ivy/test/completions_spec.ts [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+import {LanguageService} from '../language_service';
+
+import {LanguageServiceTestEnvironment} from './env';
+
+describe('completions', () => {
+  beforeEach(() => {
+    initMockFileSystem('Native');
+  });
+
+  describe('in the global scope', () => {
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:23]: Could we add a test getting for completions over no/empty node, like `{{ | }}`?

File: packages/language-service/ivy/completions.ts (hunk 1 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:19]: Rephrased, good point!

File: packages/language-service/ivy/completions.ts (hunk 4 above)
Comment (ayazhafiz) [reply packages/language-service/ivy/completions.ts:44]: I meant that you could just do
```suggestion
    if (this.isPropertyExpressionCompletion()) {
      return this.getPropertyExpressionCompletion(options);
    }
    return undefined;
```

no action required, just a note

File: packages/language-service/ivy/completions.ts (hunk 8 above)
Comment (ayazhafiz) [reply packages/language-service/ivy/completions.ts:175]: I meant, could you add a comment on why it is OK to call that function here? I know it's covered by the function jsdoc but I think it would be helpful to reiterate that it is sound to do it here because we are using it as a string.

File: packages/language-service/ivy/completions.ts (hunk 6 above)
Comment (ayazhafiz) [reply packages/language-service/ivy/completions.ts:45]: :+1:

File: packages/language-service/ivy/completions.ts (hunk 7 above)
Comment (ayazhafiz) [reply packages/language-service/ivy/completions.ts:90]: +1

File: packages/language-service/ivy/test/completions_spec.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+      const completions = ngLS.getCompletionsAtPosition(fileName, cursor, /* options */ undefined);
+      expectContain(completions, ts.ScriptElementKind.memberVariableElement, ['title', 'hero']);
+    });
+
+    it('should be able to complete a property binding', () => {
+      const {ngLS, fileName, cursor} =
+          setup('<h1 [model]="ti¦"></h1>', `title!: string; hero!: number;`);
+      const completions = ngLS.getCompletionsAtPosition(fileName, cursor, /* options */ undefined);
+      expectContain(completions, ts.ScriptElementKind.memberVariableElement, ['title', 'hero']);
+    });
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:41]: Does an empty property binding work? `[model]=""`

File: packages/language-service/ivy/test/completions_spec.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+    it('should be able to retrieve details for completions', () => {
+      const {ngLS, fileName, cursor} = setup('{{ti¦}}', `
+      /** This is the title of the 'AppCmp' Component. */
+      title!: string;
+      /** This comment should not appear in the output of this test. */
+      hero!: number;
+    `);
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:49]: nit
```suggestion
      const {ngLS, fileName, cursor} = setup('{{ti¦}}', `
        /** This is the title of the 'AppCmp' Component. */
        title!: string;
        /** This comment should not appear in the output of this test. */
        hero!: number;
      `);
```

File: packages/language-service/ivy/test/completions_spec.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+          </div>
+        `,
+          `heroes!: {name: string}[];`);
+      const completions = ngLS.getCompletionsAtPosition(fileName, cursor, /* options */ undefined);
+      expectContain(completions, ts.ScriptElementKind.memberVariableElement, ['heroes']);
+      expectContain(completions, DisplayInfoKind.VARIABLE, ['hero']);
+    });
+
+    it('should return completions inside an event binding', () => {
+      const {ngLS, fileName, cursor} = setup(` <button (click)='t¦'></button>`, `title!: string;`);
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:79]: Similarly, does an empty event binding work?

Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:79]: nit
```suggestion
      const {ngLS, fileName, cursor} = setup(`<button (click)='t¦'></button>`, `title!: string;`);
```

File: packages/language-service/ivy/test/completions_spec.ts [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+      const completions = ngLS.getCompletionsAtPosition(fileName, cursor, /* options */ undefined);
+      expectContain(completions, ts.ScriptElementKind.memberVariableElement, ['title']);
+    });
+
+    it('should return completions inside the RHS of a two-way binding', () => {
+      const {ngLS, fileName, cursor} = setup(`<h1 [(model)]="t¦"></h1>`, `title!: string;`);
```
Comment (ayazhafiz) [inline packages/language-service/ivy/test/completions_spec.ts:85]: Similarly, does an empty two-way binding work?

**Review by ayazhafiz — APPROVED** [review]
Other than my outstanding comments LGTM, this is awesome! adding Keen and Andrew as reviewers as well as they are probably interested too.

File: packages/language-service/ivy/completions.ts [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+
+  /**
+   * Get completions for a property expression in a global context (e.g. `{{y|}}`).
+   */
+  private getGlobalPropertyExpressionCompletion(
+      this: PropertyExpressionCompletionBuilder,
+      options: ts.GetCompletionsAtPositionOptions|
+      undefined): ts.WithMetadata<ts.CompletionInfo>|undefined {
+    const completions =
+        this.compiler.getTemplateTypeChecker().getGlobalCompletions(this.context, this.component);
```
Comment (atscott) [inline packages/language-service/ivy/completions.ts:138]: suggestion: maybe have a property `private readonly templateTypeChecker = this.compiler.getTemplateTypeChecker()`

File: packages/language-service/ivy/completions.ts [hunk 18]
```diff
@@ -0,0 +1,262 @@
...
+                                                       DisplayInfoKind.VARIABLE),
+      });
+    }
+
+    return {
+      entries,
+      // Although this completion is "global" in the sense of an Angular expression (there is no
+      // explicit receiver), it is not "global" in a TypeScript sense since Angular expressions have
+      // the component as an implicit receiver.
+      isGlobalCompletion: false,
```
Comment (atscott) [inline packages/language-service/ivy/completions.ts:186]: optional bikeshed (though at this point, I'm leaning towards keeping this how it is because I can't think of a better name and it'd be a pain to rename):

This made me think about how this isn't really a "global" completion, but more like a "component scope" completion.

File: packages/language-service/ivy/completions.ts [hunk 19]
```diff
@@ -0,0 +1,262 @@
...
+      return undefined;
+    }
+    const {componentContext, templateContext} = completions;
+
+    if (templateContext.has(entryName)) {
+      const entry = templateContext.get(entryName)!;
+      // Entries that reference a symbol in the template context refer either to local references or
+      // variables.
+      const symbol = templateTypeChecker.getSymbolOfNode(entry.node, this.component) as
+              TemplateDeclarationSymbol |
```
Comment (atscott) [inline packages/language-service/ivy/completions.ts:212]: Suggestion: maybe we should update the overloads of `getSymbolOfNode` to return the more narrow types for `TmplAstVariable` and `TmplAstReference` inputs?

File: packages/language-service/ivy/completions.ts (hunk 5 above)
Comment (atscott) [reply packages/language-service/ivy/completions.ts:66]: Did you forget to push up this change? I still see the `console.error`

File: packages/language-service/ivy/language_service.ts [hunk 20]
```diff
@@ -97,6 +98,70 @@ export class LanguageService {
...
+    const program = this.strategy.getProgram();
+    const compiler = this.compilerFactory.getOrCreateWithChangedFile(fileName);
+    const templateInfo = getTemplateInfoAtPosition(fileName, position, compiler);
+    if (templateInfo === undefined) {
+      return undefined;
+    }
+    const positionDetails = getTargetAtPosition(templateInfo.template, position);
+    if (positionDetails === null) {
+      return undefined;
+    }
```
Comment (atscott) [inline packages/language-service/ivy/language_service.ts:155]: Suggestion: these ~10 lines appear to be exact duplicates of each other in each of these completion functions. Is it possible to refactor to a helper method? (I think this is also pretty much the same work as all of the other "getXAtPosition" with fileName and position)

File: packages/language-service/ivy/test/completions_spec.ts [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+} {
+  const codePath = absoluteFrom('/test.ts');
+  const templatePath = absoluteFrom('/test.html');
+  const env = LanguageServiceTestEnvironment.setup([
+    {
+      name: codePath,
+      contents: `
+        import {Component, NgModule} from '@angular/core';
+
+        @Component({templateUrl: './test.html', selector: 'app-cmp',}) export class AppCmp {
```
Comment (atscott) [inline packages/language-service/ivy/test/completions_spec.ts:121]: nit: formatting here could be a bit better

File: packages/language-service/ivy/test/completions_spec.ts [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+        export class AppModule {}
+        `,
+      isRoot: true,
+    },
+    {
+      name: templatePath,
+      contents: 'Placeholder template',
+    }
+  ]);
+  const {nodes, cursor} = env.overrideTemplateWithCursor(codePath, 'AppCmp', templateWithCursor);
```
Comment (atscott) [inline packages/language-service/ivy/test/completions_spec.ts:137]: suggestion: You actually kind of don't need `overrideTemplateWithCursor` if you expose `extractCursorInfo`.
You can instead do 
```
const {text: templateText, cursor} = extractCursortInfo(templateWithCursor);
...
{
  name: templatePath,
  contents: templateText
}
```

Especially given that you don't need the overridden template `nodes`.

I kind of prefer this because it eliminates some of the possible string errors (i.e. mistyping the class name) and doesn't create the file initially with placeholder text.

**Review by atscott — APPROVED** [review]

File: packages/language-service/ivy/completions.ts [hunk 23]
```diff
@@ -0,0 +1,262 @@
...
+    if (completions === null) {
+      return undefined;
+    }
+    const {componentContext, templateContext} = completions;
+    if (templateContext.has(entryName)) {
+      const node: TmplAstReference|TmplAstVariable = templateContext.get(entryName)!.node;
+      const symbol =
+          templateTypeChecker.getSymbolOfNode(node, this.component) as TemplateDeclarationSymbol |
+          null;
+      if (symbol === null || symbol.tsSymbol === null) {
```
Comment (kyliau) [inline packages/language-service/ivy/completions.ts:252]: nit: consider replacing these 4 lines with `return symbol?.tsSymbol`. I think it is in line with the Angular "style guide" since this function could return undefined.

File: packages/language-service/ivy/language_service.ts [hunk 24]
```diff
@@ -97,6 +98,70 @@ export class LanguageService {
...
+
+    const result = new CompletionBuilder(
+                       this.tsLS, compiler, templateInfo.component, positionDetails.node,
+                       positionDetails.context)
+                       .getCompletionEntryDetails(entryName, formatOptions, preferences);
+    this.compilerFactory.registerLastKnownProgram();
+    return result;
+  }
+
+  getCompletionEntrySymbol(fileName: string, position: number, name: string): ts.Symbol|undefined {
```
Comment (kyliau) [inline packages/language-service/ivy/language_service.ts:145]: Could you please add a comment to describe the difference between `getCompletionEntryDetails` and `getCompletionEntrySymbol`? When is one used over the other?

**Review by kyliau — APPROVED** [review]

File: packages/language-service/ivy/language_service.ts (hunk 24 above)
Comment (alxhub) [reply packages/language-service/ivy/language_service.ts:153]: Oh I have no idea - these are both defined in the TypeScript API so we implement them both.

File: packages/language-service/ivy/completions.ts (hunk 8 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:175]: I don't think that's helpful or reasonable to do - this function is called all over the place in the Language Service code, with the exact same rationale - it's because the TS API doesn't allow us to express custom types unless we "force" them into the string enum that it expects. I don't think that duplicating that explanation in dozens of places in our codebase will provide much benefit.

File: packages/language-service/ivy/completions.ts (hunk 18 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:186]: This is a TypeScript-defined field in `ts.CompletionInfo`, we have no control over the name.

Comment (atscott) [reply packages/language-service/ivy/completions.ts:186]: Right, what I meant was to possibly rename the entire implementation here to not say "global" because a "global" completion, as defined by this typescript field, is something else. But I'm also fine leaving this as-is.

File: packages/language-service/ivy/test/completions_spec.ts (hunk 13 above)
Comment (alxhub) [reply packages/language-service/ivy/test/completions_spec.ts:41]: Added a test

File: packages/language-service/ivy/completions.ts (hunk 5 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:66]: Yeah, I think I hadn't pushed the latest.

File: packages/language-service/ivy/test/completions_spec.ts (hunk 15 above)
Comment (alxhub) [reply packages/language-service/ivy/test/completions_spec.ts:79]: It doesn't, currently, due to the way the template parser produces a weird, invalid token. I added a excluded test, which will be enabled after my refactoring of `template_target`.

File: packages/language-service/ivy/language_service.ts (hunk 24 above)
Comment (kyliau) [reply packages/language-service/ivy/language_service.ts:153]: Oh, okay. For our purpose I don't think this matters much since we pick the API to call in the language server, but I'll add a card for myself to investigate how they're used in TypeScript language service.

Comment (ayazhafiz) [reply packages/language-service/ivy/language_service.ts:153]: `getCompletionEntrySymbol` was added because of https://github.com/microsoft/TypeScript/issues/10812, but the issue doesn't give an example motivation.

File: packages/language-service/ivy/test/completions_spec.ts (hunk 16 above)
Comment (alxhub) [reply packages/language-service/ivy/test/completions_spec.ts:85]: It took a bit of fixing, but I got this working, due to the broken way that such bindings are parsed.

File: packages/language-service/ivy/completions.ts (hunk 18 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:186]: Ahhh, yeah, I see what you mean.

File: packages/language-service/ivy/completions.ts (hunk 19 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:212]: I don't think that would actually work in this circumstance, as TypeScript will match a signature based on the union input type (`TmplAstVariable|TmplAstReference`), and not "split" the union to match each branch individually and then create a union of the return types. So we'll need the cast either way.

File: packages/language-service/ivy/completions.ts (hunk 8 above)
Comment (ayazhafiz) [reply packages/language-service/ivy/completions.ts:175]: 👍

File: packages/language-service/ivy/language_service.ts (hunk 20 above)
Comment (alxhub) [reply packages/language-service/ivy/language_service.ts:163]: Done - `getCompletionBuilder` is now a thing.

File: packages/language-service/ivy/test/completions_spec.ts (hunk 22 above)
Comment (alxhub) [reply packages/language-service/ivy/test/completions_spec.ts:137]: 👍 but will leave this as a future refactoring. I'd prefer to find a way not to "decouple" the `extractCursorInfo` operation from the path/content it's representing.

File: packages/language-service/ivy/completions.ts (hunk 23 above)
Comment (alxhub) [reply packages/language-service/ivy/completions.ts:252]: I don't think that would work, since `symbol.tsSymbol` would be `null` and not `undefined`. You'd need `symbol?.tsSymbol ?? undefined`. I prefer the explicitness of the `if` branch, which makes it very clear that this is exactly how we want to handle both cases.
